Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp

```diff
@@ -559,8 +664,104 @@ bool Connection::sendOutputMessage(UnixMessage& outputMessage)
         return false;
     }
 
+#if OS(ANDROID)
+    return sendHardwareBuffers(WTFMove(hardwareBuffers));
+#else
+    return true;
+#endif
+}
+
+#if OS(ANDROID)
+bool Connection::sendHardwareBuffers(Deque<RefPtr<AHardwareBuffer>, 2>&& hardwareBuffers)
+{
+    RELEASE_ASSERT(m_outgoingHardwareBuffers.isEmpty());
+    m_outgoingHardwareBuffers = WTFMove(hardwareBuffers);
+    return sendOutgoingHardwareBuffers();
+}
+
+bool Connection::sendOutgoingHardwareBuffers()
+{
+    while (!m_outgoingHardwareBuffers.isEmpty()) {
+        auto& buffer = m_outgoingHardwareBuffers.first();
+        RELEASE_ASSERT(buffer);
+
+        // There is no need to check for EINTR, it is handled internally.
+        int result = AHardwareBuffer_sendHandleToUnixSocket(buffer.get(), socketDescriptor());
+        if (!result) {
+            m_outgoingHardwareBuffers.removeFirst();
+            continue;
+        }
+
+        if (result == -EAGAIN || result == -EWOULDBLOCK) {
+            m_writeSocketMonitor.start(m_socket.get(), G_IO_OUT, m_connectionQueue->runLoop(), [this, protectedThis = Ref { *this }] (GIOCondition condition) -> gboolean {
+                if (condition & G_IO_OUT) {
+                    RELEASE_ASSERT(!m_outgoingHardwareBuffers.isEmpty());
+                    // We can't stop the monitor from this lambda, because stop destroys the lambda.
+                    m_connectionQueue->dispatch([this, protectedThis = Ref { *this }] {
+                        m_writeSocketMonitor.stop();
+                        if (m_isConnected) {
+                            if (sendOutgoingHardwareBuffers())
+                                sendOutgoingMessages();
+                        }
+                    });
+                }
+                return G_SOURCE_REMOVE;
+            });
+            return false;
+        }
+
+        if (result == -EPIPE || result == -ECONNRESET) {
+            connectionDidClose();
+            return false;
+        }
+
+        if (m_isConnected) {
+            LOG_ERROR("Error sending AHardwareBuffer on socket %d in process %d: %s", socketDescriptor(), getpid(), safeStrerror(-result).data());
+            connectionDidClose();
+        }
+        return false;
+    }
+
+    RELEASE_ASSERT(m_outgoingHardwareBuffers.isEmpty());
+    return true;
+}
+
+bool Connection::receiveHardwareBuffers(size_t bufferCount)
+{
+    RELEASE_ASSERT(!m_pendingIncomingHardwareBufferCount);
+    RELEASE_ASSERT(m_incomingHardwareBuffers.isEmpty());
+
+    m_pendingIncomingHardwareBufferCount = bufferCount;
+    return receiveIncomingHardwareBuffers();
+}
```

Comment:
Same here

Reply:
Same reason as above: one function does setup and checks, the other one the actual work. I'll move it, too.